### PR TITLE
Use `npm pack pkg@version` to download tarball

### PR DESCRIPTION
**Pull request by qlonik**
I have attempted to use this package with private npm registry. With that registry I was having the following issue `bower angular-dw-auth#0.1.0  ENOTEMPTY ENOTEMPTY: directory not empty, rmdir 'C:\Users\user\.babun\cygwin\tmp\tmp-110188n13rawkgAFuB'`

Said registry also requires authentication information to fetch the attachement. I have not found in the download.js `fetch` method any mentions of passing username and password or token. So, instead of parsing npm config for information about login information, this method uses npm itself to download the tarball. This method should delegate managing of proxy/authentication/fetch to npm. This way different enpoints for different scopes should also be delegated to npm, so npm will use one host for one scope and another host for another scope, all based on the global npm settings.

### CODE REVIEW

**Comment by mjeanroy** [comment]
Thanks @qlonik!
I did not know that `npm pack` command could be used to download a tarball. It looks like this is the right thing to do (and it should fix #1 by the way).

It seems that tests are failing, can you check it if you have some time?

File: src/npm-utils.js [hunk 1]
```diff
@@ -56,6 +58,20 @@ var execViewCommand = function(args) {
   return deferred.promise;
 };
 
+var execPackCommand = function(args) {
```
Comment (jakub-g) [inline src/npm-utils.js]: Would be nice to put some jsdoc on this method (params, return type) ;) probably a link to npm docs about what `args` is

**Comment by qlonik** [comment]
Sure. The tests are failing right now because I removed download.js and some other functions, but didnt remove respective tests. There should couple tests be written for the new function to download tarball as well.

**Comment by qlonik** [comment]
I fixed tests and added a little bit more changes to downloadTarball function.
If this one is good, could it be merged into the project?

File: src/npm-utils.js [hunk 2]
```diff
@@ -121,33 +119,24 @@ module.exports = {
...
-
-        // Otherwise, unwrap it.
-        var mostRecentVersion = getLastKey(data);
-        return data[mostRecentVersion]['dist.tarball'];
+  downloadTarball: function(pkg, version, dir) {
+    var oldCWD = process.cwd();
+    process.chdir(dir);
+    return execPackCommand([pkg + '@' + version])
+      .then(function(filename) {
+        process.chdir(oldCWD);
```
Comment (mjeanroy) [inline src/npm-utils.js]: Can you move this line in the `finally` callback of the promise (to reset the process working directory even if something failed with `execPackCommand`)?

File: test/npm-utils-spec.js [hunk 3]
```diff
@@ -25,33 +25,32 @@
...
       })
       .finally(function() {
         done();
       });
   });
 
-  it('should get tarball url', function(done) {
-    var promise = npmUtil.tarball('bower', '1.7.7');
+  it('should get tarball', function(done) {
+    var promise = npmUtil.downloadTarball('bower', '1.7.7', __dirname);
```
Comment (mjeanroy) [inline test/npm-utils-spec.js]: Can you download the tarball to a temporary directory instead of `__dirname` (directory that will be removed after test)?

**Comment by mjeanroy** [comment]
@qlonik That seems great, thank you! I have two (small) comments on your PR, once done I will merge it and, probably, release a new version!

**Comment by mjeanroy** [comment]
Thanks @qlonik!

**Comment by mjeanroy** [comment]
Hi, v0.4.0 (should be available on npm) is released and use `npm pack` to download package, thanks @qlonik!

**Comment by qlonik** [comment]
That's awesome! Thanks for pulling this one in.

**Pull request merged**